logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,240 @@
+   *
+   * Close the server socket, shutdown the thread pool, stop accepting new requests,
+   * and blocks until all worker threads terminate.
+   */
+  private void stopServing() {
+    mStopped = true;
+    if (mServerSocket != null) {
+      try {
+        mServerSocket.close();
+      } catch (IOException e) {
+        LOG.warn("Exception in closing server socket.", e);
+      }
+    }
+
+    mThreadPool.shutdown();
+    long timeoutMS = STOP_TIMEOUT_MS;
+    long now = System.currentTimeMillis();
+    while (timeoutMS >= 0) {
+      try {
+        boolean ret = mThreadPool.awaitTermination(timeoutMS, TimeUnit.MILLISECONDS);
+        if (ret) {
+          LOG.info("All worker threads have terminated.");
+        } else {
+          LOG.warn("Log server has timeout waiting for worker threads to terminate.");
+        }
+        break;
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        long newnow = System.currentTimeMillis();
+        timeoutMS -= (newnow - now);
+        now = newnow;
+      }
+    }
+  }
+
+  /**
+   * Configure a {@link Hierarchy} instance used to retrive logger by name and maintain the logger
+   * hierarchy. An instance of this class will be passed to a {@link AlluxioLog4jSocketNode} so
+   * that the {@link AlluxioLog4jSocketNode} instance can retrieve the logger to log incoming
+   * {@link org.apache.log4j.spi.LoggingEvent}s.
+   *
+   * @param inetAddress inet address of the client
+   * @param logAppenderName name of the appender to use for this client
+   * @return a {@link Hierarchy} instance to pass to {@link AlluxioLog4jSocketNode}
+   * @throws IOException if fails to create an {@link FileInputStream} to read log4j.properties
+   */
+  protected LoggerRepository configureHierarchy(InetAddress inetAddress, String logAppenderName)
+      throws IOException {
+    Hierarchy clientHierarchy;
+    String inetAddressStr = inetAddress.getHostAddress();
+    Properties properties = new Properties();
+    File configFile;
+    try {
+      configFile = new File(new URI(System.getProperty("log4j.configuration")));
+    } catch (URISyntaxException e) {
+      // Alluxio log server cannot derive a valid path to log4j.properties. Since this
+      // properties file is global, we should throw an exception.
+      LOG.error("Cannot derive a valid URI to log4j.properties file.");
+      throw new RuntimeException(e);
+    }
+    try (FileInputStream inputStream = new FileInputStream(configFile)) {
+      properties.load(inputStream);
+    }
+    Level level = Level.INFO;
+    clientHierarchy = new Hierarchy(new RootLogger(level));
+    // Startup script should guarantee that mBaseLogsDir already exists.
+    String logDirectoryPath = mBaseLogsDir + "/" + logAppenderName.toLowerCase();
+    File logDirectory = new File(logDirectoryPath);
+    if (!logDirectory.exists()) {
+      logDirectory.mkdir();
+    }
+    String logFilePath = logDirectoryPath + "/" + inetAddressStr + ".log";
```

Comment:
Will we automatically handle rolling over the log file?

Reply:
Yes, we are using log4j's RollingFileAppender.